Request_NOAA

ssolson opened the pull request:
Adds functionality to request historical wave spectral data from NOAA.

There are 3 new functions:
`fetch_NDBC`
`_get_links`
`noaa_dates_to_DateTime`

These have been adapted from WDRT following heavy modification. The work flow is to get links which returns a dictionary of years and links to data. Then fetch_NDBC will return the yearly data as a dataframe for each year in a dictionary. Lastly `noaa_dates_to_DateTime` will convert the NOAA dates to a consistent datetime format.

My example script is as follows:

```
from mhkit.wave.io import fetch_NDBC, prep_data, noaa_dates_to_DateTime
import matplotlib.pyplot as plt

# Get a dictionary of Historical Spectral Wave Data by year
noaa_data_years = fetch_NDBC('46022')

# Create a Datetime Index remove NOOA date columns for each year
for year in noaa_data_years:
    print(year)
    year_data = noaa_data_years[year]
    year_data_in_datetime = noaa_dates_to_DateTime(year_data)
    year_data_in_datetime = year_data_in_datetime.set_index('date')
    noaa_data_years[year] = year_data_in_datetime


```

ssolson commented:
Thanks Rebecca.

I'll check out 1. 
For 2: The parameter exists to set up the framework to allow us to add another parameter following this PR. E.g. when I add one more parameter there will be significantly less work because I have already built the primary infrastructure.
For 3 I have tried my best to make the headers explain what they do. I would appreciate specific suggestions and edits which clarify the points you found confusing.

I will check the example script I provided. I am not sure what you mean by more clear examples. Do you mean one that works :)?

rpauly18 commented:
> Thanks Rebecca.
> 
> I'll check out 1.
> For 2: The parameter exists to set up the framework to allow us to add another parameter following this PR. E.g. when I add one more parameter there will be significantly less work because I have already built the primary infrastructure.
> For 3 I have tried my best to make the headers explain what they do. I would appreciate specific suggestions and edits which clarify the points you found confusing.
> 
> I will check the example script I provided. I am not sure what you mean by more clear examples. Do you mean one that works :)?

Hi Sterling, 

Yes, an example that works would be nice. It looks like you just left the first step out, which I eventually figured out. I am about to submit some changes to your PR that fix a couple errors and hopefully add more descriptive hearer info.

rpauly18 commented:
@ssolson 

I submitted commits that fix 1 and 3 (I think). As for #2, I added some text trying to explain that more parameters are coming soon. Please review and make sure I captured what the parameters are adequately. 

I am going to make similar changes to the other functions.

ssolson commented on mhkit/wave/io.py:
```diff
@@ -112,30 +112,32 @@ def read_NDBC_file(file_name, missing_values=['MM',9999,999,99]):
-    parameter: string
-        'swden'	:	'Raw Spectral Wave Current Year Historical Data'
-    number: string
+    parameter: string (optional)
```
As it currently functions this appears optional but this parameter is used in the request address. I would prefer to not list this as optional for this reason as well as there are currently no other data options available.

ssolson commented on mhkit/wave/io.py:
```diff
@@ -112,30 +112,32 @@ def read_NDBC_file(file_name, missing_values=['MM',9999,999,99]):
-        'swden'	:	'Raw Spectral Wave Current Year Historical Data'
-    number: string
+    parameter: string (optional)
+        Current option is 'swden'	:	'Raw Spectral Wave Current Year Historical Data'
```
This clutters the docstring in my opinion. I think this can go in the general description e.g. where you added `that contain that parameter data` we could either adjust that to say spectral wave density data. Or add another sentence that clarifies that `swden` is currently the only parameter supported.

ssolson commented on mhkit/wave/io.py:
```diff
@@ -112,30 +112,32 @@ def read_NDBC_file(file_name, missing_values=['MM',9999,999,99]):
-    number: string
+    parameter: string (optional)
+        Current option is 'swden'	:	'Raw Spectral Wave Current Year Historical Data'
+        More parameter options will be available soon. 
```
same as the previous comment

rpauly18 replied:
@ssolson my response to all of these comments is to ask how long you think it will be until you have the capability to add more parameter options? If this is something that can be accomplished fairly soon (month to 3 months), I'd say let's hold off releasing this code until more options are available. If you want to release this code, I would advise temporarily removing the "parameter" input and just have 'swden' hard coded at the top of the function. It would not be too heavy of a lift to re-introduce "parameter" as an input option down the road.

ssolson commented:
@rpauly18  I have added a second parameter.

kmruehl commented on mhkit/wave/io.py:
```diff
@@ -105,3 +109,303 @@ def read_NDBC_file(file_name, missing_values=['MM',9999,999,99]):
+    -------
+    ndbc_data: dict
+        Dictionary of NDBC data 
+    '''
```
Okay, so I played with this a little bit, and it's pretty simple to convert one or two years of data to a DataFrame, but what about all  historical data? I think it would be useful to make a DataFrame containing all historical data the default.

``` python
df1 = pd.DataFrame.from_dict(ndbc_data['2019'])
df2 = pd.DataFrame.from_dict(ndbc_data['2018'])
new =pd.concat([df1,df2])
```

then you should be able to run functions from `mhkit.wave.resource`
```python
mhkit.wave.resource.significant_wave_height(new['2019-01-01 00:00:00'],frequency_bins=new.columns.values)
```
but I'm getting a datetime error
"Invalid comparison between dtype=datetime64[ns] and int"

kmruehl replied:
```
Create DataFrame for all May data, plot May data
may_data = new[new.index.month==5]
may_data.mean().plot()
may_data[:5].T.plot()

Plot all data from first day in May
may_data[may_data.index.day==1].T.plot()
```

ssolson replied:
> Why not return `ndbc_data` as dictionary instead of a DataFrame? I think this return should be the same as `read_NDBC_file`?
> 
> I re-read the description and now I see that
> "`fetch_ndbc`: Takes filenames and returns a dictionary of DataFrames indexed by years for each filename passed"
> 
> What is the reasoning behind this? Why not just output a DataFrame that includes all data?
> 
> ```
> ndbc_data['2019'].head()
> ndbc_data['2001'].plot()
> ```
> 
> It's simple to combine years of data to a DataFrame, but what about all historical data? I think it would be useful to make a DataFrame containing all historical data the default.
> 
> ```python
> new = pd.concat([ndbc_data['2018'],ndbc_data['2019']])
> ```
> 
> then you should be able to run functions from `mhkit.wave.resource`
> 
> ```python
> mhkit.wave.resource.significant_wave_height(new['2019-01-01 00:00:00'],frequency_bins=new.columns.values)
> ```
> 
> but I'm getting a datetime error
> "Invalid comparison between dtype=datetime64[ns] and int"

Kelly as we discussed in our meeting last week it would be nice to have all of the data as one object. However the frequency bins and range change over the years. Combining these would lead to Nans throughout the single dataframe and would be quite messy (e.g  `new = pd.concat([ndbc_data['1996'],ndbc_data['2019']])`).  

In your next step, I think having the inconsistent frequency bins gets even messier as you specify the frequency_bins to be the column values. To use the Hs function you would currently need to transpose the data. I cannot reproduce your error message.

ssolson replied:
> Create DataFrame for all May data, plot May data
> 
> ```python
> may_data = new[new.index.month==5]
> may_data.mean().plot()
> may_data[:5].T.plot()
> ```
> 
> Plot all data from first day in May
> 
> ```python
> may_data[may_data.index.day==1].T.plot()
> ```
> 
> Resample data on annual, monthly or weekly basis
> 
> ```python
> new.resample('AS').mean().head()
> new.resample('MS').mean().head()
> new.resample('W').mean().head()
> ```

Yes having the index as a Datetime for historical data allows for this sort of analysis. I assume you want to include this as an example? Do you want to continue the discussion of bins vs time as the index. I like that the request historical data returns data as a time series. I think we could add a function to convert historical to MHKiT... or I could leave it as is and bring it up with Kate

kmruehl replied:
> Kelly as we discussed in our meeting last week it would be nice to have all of the data as one object. However the frequency bins and range change over the years. Combining these would lead to Nans throughout the single dataframe and would be quite messy (e.g `new = pd.concat([ndbc_data['1996'],ndbc_data['2019']])`).

I figured that was the motivation. How often do the frequency bins and ranges change (e.g. does it happen once due to a NDBC change, or is it somewhat random)? Would it be useful to combine all data with the same frequency bins and range into one DataFrame (even if it's not the entire data history)? Or, is it better to keep the output like this for now, and develop a separate way to merge historical data with different frequency bins and ranges?

kmruehl replied:
>  I assume you want to include this as an example? Do you want to continue the discussion of bins vs time as the index. I like that the request historical data returns data as a time series. I think we could add a function to convert historical to MHKiT... or I could leave it as is and bring it up with Kate

Yes, my thought was to include using pandas to view/analyze historical data as an example. This could be a new section or appendix do an existing example, or a new example. And I'm open to revisiting the conversation about indexing and/or adding a new function.

ssolson replied:
> > Kelly as we discussed in our meeting last week it would be nice to have all of the data as one object. However the frequency bins and range change over the years. Combining these would lead to Nans throughout the single dataframe and would be quite messy (e.g `new = pd.concat([ndbc_data['1996'],ndbc_data['2019']])`).
> 
> I figured that was the motivation. How often do the frequency bins and ranges change (e.g. does it happen once due to a NDBC change, or is it somewhat random)? Would it be useful to combine all data with the same frequency bins and range into one DataFrame (even if it's not the entire data history)? Or, is it better to keep the output like this for now, and develop a separate way to merge historical data with different frequency bins and ranges?

So as far as I have seen the bins change once (e.g. 2 distinct frequency binning). However, the date format changes twice (3 distinct date formats).  So if you wanted to combine across years on the subset of evenly spaced bins I see this as a separate function. The user would first need to have the data and convert the index to datetime and drop the original ndbc date columns (steps 2 & 3 as covered in the example). I have already given a number of reasons why I think these should be separate steps (summary calculations are separate from requests, should we manipulate the underlying dataframe, what if the user has custom functions which expect the data in the ndbc reported format?).  Then the user could combine the subset of evenly spaced frequency bins but then the naming of the dataframe is weird for the dictionary of dataframes E.g. key value would now be '1996-2004' and '2004b-2019'. I don't see writing this to be a built-in function as explicitly useful. Where as I do think the next step would be to determine the best method to combine uneven binning and ranges and add a function for that so that we would have one object.

ssolson replied:
> > I assume you want to include this as an example? Do you want to continue the discussion of bins vs time as the index. I like that the request historical data returns data as a time series. I think we could add a function to convert historical to MHKiT... or I could leave it as is and bring it up with Kate
> 
> Yes, my thought was to include using pandas to view/analyze historical data as an example. This could be a new section or appendix do an existing example, or a new example. And I'm open to revisiting the conversation about indexing and/or adding a new function.

Okay I will keep this function to return Datetime and once we reach an agreement bring it up with Kate

kmruehl replied:
> In your next step, I think having the inconsistent frequency bins gets even messier as you specify the frequency_bins to be the column values. To use the Hs function you would currently need to transpose the data. I cannot reproduce your error message.

Thanks for the reminder about indexing by frequency, not DateTime. When I transpose the data I get a different error
``` python
new = pd.concat([ndbc_data['2018'],ndbc_data['2019']])
mhkit.wave.resource.significant_wave_height(new.T)
```
_TypeError: '>' not supported between instances of 'str' and 'int'_


Then I tried specifying the frequency spacing, I get the same error
``` python
new = pd.concat([ndbc_data['2018'],ndbc_data['2019']])
 mhkit.wave.resource.significant_wave_height(new.T,frequency_bins=new.columns.values)
```
_TypeError: '>' not supported between instances of 'str' and 'int'_

ssolson replied:
So the columns are strings of frequency bins. you would need to transpose and covert the index I believe.

ssolson commented on mhkit/wave/io.py:
```diff
@@ -105,3 +109,303 @@ def read_NDBC_file(file_name, missing_values=['MM',9999,999,99]):
+    return available_data
+    
+def _ndbc_parse_filenames(parameter, filenames):  
+    '''
```
ssolson@DESKTOP-E1C7N17:/mnt/c/Users/sterl/Codes/MHKiT-Python/mhkit$ grep -R def\ _ .
./river/graphics.py:def _xy_plot(x, y, fmt='.', label=None, xlabel=None, ylabel=None, title=None,
./river/io.py:def _read_usgs_json(text):
./tidal/graphics.py:def _initialize_polar(ax = None, metadata=None, flood=None, ebb=None):
./tidal/io.py:def _json_to_dataframe(response):
./tidal/io.py:def _xml_to_dataframe(response):
./tidal/resource.py:def _histogram(directions, velocities, width_dir, width_vel):
./tidal/resource.py:def _normalize_angle(degree):
./wave/io.py:def _ndbc_parse_filenames(parameter, filenames):
./wave/io.py:def _date_string_to_datetime(df, columns, year_fmt):
./wave/io.py:def __supported_ndbc_params(parameter):
./wave/performance.py:def _std_ddof1(a):
./wave/performance.py:def _performance_matrix(X, Y, Z, statistic, x_centers, y_centers):
./wave/performance.py:    def _frequency(a):

kmruehl replied:
Thanks @ssolson, I was still drafting this PR review and hadn't checked the source code yet. My recollection was that private methods were internal to public methods, or had been removed, but that's not the case.

ssolson commented on mhkit/wave/io.py:
```diff
@@ -105,3 +109,303 @@ def read_NDBC_file(file_name, missing_values=['MM',9999,999,99]):
+        'swden'	:	'Raw Spectral Wave Current Year Historical Data'
+        'stdmet':   'Standard Meteorological Current Year Historical Data'
+    buoy_number: string (optional)
+        Buoy Number.  5-character alpha-numeric station identifier        
```
Good idea. Added in 425edd274fd16383328108d7f8eb6075f35169ee

kmruehl replied:
Thanks

ssolson commented:
> **this is a draft**
> I have several comments and suggested revisions, but they are not complete yet,
> 
> * method names should be consistent with ~~river~~ wave.io module, e.g. `read_NDBC_file` and `ndbc_available_data` and `ndbc_available_data`
Not a method (same for all below). As stated in our meeting I believe we should rewrite 'read_ndbc_file' following the approval of this PR to use the datetime utils. 
> * method names should be consistent within ~~wave.io~~ river.io e.g. `read_usgs_file` and `request_usgs_data`
I think they are. Please provide more detail.
> * method output should be a DataFrame for consistency and to run functions from `mhkit.wave.resource`
 And see discussion about bins over the years.
> * `ndbc_dates_to_datetime` should be moved to utils, https://mhkit-software.github.io/MHKiT/mhkit-python/api.utils.html, and having similar naming, i.e.  `ndbc_to_datetime`
> * Why are these split into separate methods instead of combined into one, or two?
Steps 1 and 2 are split out so that one may determine first what data/ years are available and the **down select** from that data. E.g most people do not know what years are available for a given buoy so forcing them to specify a range does not make sense to me and further forcing them to get all the data when they may not want the data seems wasteful. Further, the `ndbc_available_data` function if not given a buoy number returns all buoys which a useful list to know what is available to pull from in terms of buoys/ years. 

Steps 2 and 3 are separate because it is possible that someone may wish to use the native NDBC date format (I also believe that these are inherently different as one is getting data from a server to use and the other is performing a calculation using the requested data. As a functional-based toolbox, I see these as distinct operations.).


> * Why not build this into the method
> 
> ```
> for year in ndbc_data:
>    print(year)
>    year_data = ndbc_data[year]
>    year_data['date'], ndbc_date_cols = ndbc_dates_to_datetime(parameter, year_data, return_date_cols=True)
>    year_data = year_data.drop(ndbc_date_cols, axis=1)
>    year_data = year_data.set_index('date')
>    ndbc_data[year] = year_data
> ```
A method is a function that belongs to an object. As a functional-based toolbox, this is not built into the function because I am manipulating the dataframe. in a number of different ways. In this example I do not believe it is a given that the user wants to drop what I call the `ndbc_date_cols` and this is an optional return from the function. Secondly, I do not believe we should edit the index of a dataframe. The function I wrote returns a series based off a single dataframe that the user may use as they see fit. E.g. I show 1 example of a  way I think people may choose to use the `ndbc_dates_to_datetime`. Trying to make this work for dictionaries, dataframes, optionally remove columns, optional set index, and optionally overwrite data creates an unnecessarily complicated function. `ndbc_dates_to_datetime` currently does exactly what is says and adding more options only complicates.

ssolson commented:
@kmruehl @rpauly18 From 3 of Kelly's comments 
1. To move the `ndbc_date_to_datetime()` into utils, 
2. To add functionality to return the parameter data units (`ndbc_parameter_units()`) 
3. To be consistent across the naming of functions (all functions start with ndbc (`ndbc_*`)) 

I have decided to create an `ndbc` submodule as part of `wave.io`. This is because I had already renamed all the functions to start with ndbc e.g. `ndbc_*`. Adding the functionality to return parameter units brought the total new ndbc functions created to 2 ndbc functions in `io` and 2 ndbc functions in `utils`. I think grouping all these functions into an ndbc submodule of io made discovering the functions more accessible.   I have updated my example script at the top. Please let me know if you guys think we should keep the previous file structure keeping in mind this will make even more sense when I add CDIP functionality.

kmruehl commented:
Thanks @ssolson! Is this ready for final review? It seems like 7 and 9 will be addressed separate to this PR.

> I have decided to create an `ndbc` submodule as part of `wave.io`. This is because I had already renamed all the functions to start with ndbc e.g. `ndbc_*`. Adding the functionality to return parameter units brought the total new ndbc functions created to 2 ndbc functions in `io` and 2 ndbc functions in `utils`. I think grouping all these functions into an ndbc submodule of io made discovering the functions more accessible. I have updated my example script at the top. Please let me know if you guys think we should keep the previous file structure keeping in mind this will make even more sense when I add CDIP functionality.

I think this is the right choice, it makes the NDBC functions easy to find, and there really is no reason to use them unless you are specifically working with NDBC data.

ssolson commented:
> Thanks @ssolson! Is this ready for final review? It seems like 7 and 9 will be addressed separate to this PR.
> 
> > I have decided to create an `ndbc` submodule as part of `wave.io`. This is because I had already renamed all the functions to start with ndbc e.g. `ndbc_*`. Adding the functionality to return parameter units brought the total new ndbc functions created to 2 ndbc functions in `io` and 2 ndbc functions in `utils`. I think grouping all these functions into an ndbc submodule of io made discovering the functions more accessible. I have updated my example script at the top. Please let me know if you guys think we should keep the previous file structure keeping in mind this will make even more sense when I add CDIP functionality.
> 
> I think this is the right choice, it makes the NDBC functions easy to find, and there really is no reason to use them unless you are specifically working with NDBC data. And, as you mentioned, we will have a similar set of functions for CDIP.

Kelly, I am finishing up with one last adjustment. Because I added the ability to get multiple buoys from `ndbc.available_data` if a user were to pass that the filenames into `ndbc.request_data` the dictionary would overwrite the years of the buoys. So I have made another layer to the dictionary. E.g. a user will now call `ndbc_data['46022']['2014']` to get buoy 46022 year 2014 Dataframe` . I am open to other solutions.

kmruehl commented:
> Kelly, I am finishing up with one last adjustment. Because I added the ability to get multiple buoys from `ndbc.available_data` if a user were to pass that the filenames into `ndbc.request_data` the dictionary would overwrite the years of the buoys. So I have made another layer to the dictionary. E.g. a user will now call `ndbc_data['46022']['2014']` to get buoy 46022 year 2014 Dataframe` . I am open to other solutions.

That makes sense. My only suggestion is check if there multiple buoy inputs, and if so use this structure. If not, I would keep it as it was before because I don't think it makes sense to specify the buoy if there is only data for one buoy. But maybe it's better to just have the same formatting no matter what? I can see pros/cons both ways.

ssolson commented:
> > Kelly, I am finishing up with one last adjustment. Because I added the ability to get multiple buoys from `ndbc.available_data` if a user were to pass that the filenames into `ndbc.request_data` the dictionary would overwrite the years of the buoys. So I have made another layer to the dictionary. E.g. a user will now call `ndbc_data['46022']['2014']` to get buoy 46022 year 2014 Dataframe` . I am open to other solutions.
> 
> That makes sense. My only suggestion is check if there multiple buoy inputs, and if so use this structure. If not, I would keep it as it was before because I don't think it makes sense to specify the buoy if there is only data for one buoy. But maybe it's better to just have the same formatting no matter what? I can see pros/cons both ways.

I implemented the returned dictionary with just year for a single buoy and buoy & year for multiple buoys. A multi buoy example will now work in the provided example by changing the available_data() line:

```
...
available_data= ndbc.available_data(parameter, buoy_number=['46022', '44020'])
filenames= available_data['filename']
ndbc_data = ndbc.request_data(parameter, filenames)
```

kmruehl commented on mhkit/wave/io/ndbc.py:
```diff
@@ -0,0 +1,618 @@
+        Buoy Number.  5-character alpha-numeric station identifier        
+    proxy: string (optional)
+        proxy url
+        
```
@ssolson I have one more minor request, can you specify the format for the proxy setting? For example
`'http://proxy_server.com:port_number'`

ssolson replied:
> @ssolson I have one more minor request, can you specify the format for the proxy setting? For example
> `'http://proxy_server.com:port_number'`

Request 2.0 made proxies dictionaries. This should be fixed in 
2f0c97b. Let me know if you are still having issues.

kmruehl commented:
Proxy issues have now been resolved, thanks @ssolson
